File: scripts/_lib/load.ts
/**
 * Shared data loaders for bun scripts.
 *
 * Loads the trimmed JSON files in `data/` and rehydrates to the full
 * runtime shape (with derived candidate.alliance / status / margin /
 * isNota) so analysis scripts can iterate per-candidate without
 * recomputing the same logic 18 times.
 *
 * Mirrors `src/lib/data/constituencies.ts:hydrateConstituency` — kept
 * in sync by hand. The src/ runtime is the source of truth; scripts
 * are batch consumers.
 */
import { readFileSync } from "node:fs"
import { join, dirname } from "node:path"
import { fileURLToPath } from "node:url"

const ROOT = join(dirname(fileURLToPath(import.meta.url)), "..", "..")

// ── Types ────────────────────────────────────────────────────────────

export type AllianceCode = "UDF" | "LDF" | "NDA" | "OTHER" | "NOTA"

export type Candidate2026 = {
  name: string
  party: string
  alliance: AllianceCode
  votes: number
  margin: number
  status: "won" | "lost" | "nota"
  isNota: boolean
}

export type Constituency2026 = {
  constituencyNumber: number
  constituencyName: string
  candidates: Candidate2026[]
}

export type HistoricalCandidate = {
  name: string
  party: string
  alliance: AllianceCode
  votes: number
  votePct: number
}

export type HistoricalElection = {
  year: number
  type: "general" | "by-election"
  reason: string | null
  candidates: HistoricalCandidate[]
  margin: number | null
  marginPct: number | null
  turnout: number | null
  turnoutPct: number | null
  result: string | null
}

export type HistoricalConstituency = {
  constituencyNumber: number
  constituencyName: string
  elections: HistoricalElection[]
}

export type AlliancesMeta = {
  alliances: Record<
    AllianceCode,
    {
      code: AllianceCode
      name: string
      color: string
      ledBy: string
    }
  >
  partyToAlliance: Record<string, AllianceCode>
  partyAbbreviation: Record<string, string>
  partyAliases: Record<string, string>
}

// ── Loaders ──────────────────────────────────────────────────────────

function readJSON<T>(rel: string): T {
  return JSON.parse(readFileSync(join(ROOT, rel), "utf8")) as T
}

export function loadAlliances(): AlliancesMeta {
  return readJSON<AlliancesMeta>("data/alliances.json")
}

export function loadAcNames(): Record<string, { primary: string; eci: string }> {
  type Raw = Record<string, { primary: string }>
  const raw = readJSON<Raw>("data/ac-names.json")
  const out: Record<string, { primary: string; eci: string }> = {}
  for (const [k, v] of Object.entries(raw)) {
    out[k] = { primary: v.primary, eci: v.primary.toUpperCase() }
  }
  return out
}

/**
 * Load + hydrate 2026 results. Returns the array in AC order.
 * Per-candidate alliance/status/margin/isNota are derived (since the
 * source JSON now stores only name/party/votes — see commit ac1d440).
 */
export function load2026(): Constituency2026[] {
  type RawCandidate = { name: string; party: string; votes: number }
  type RawFile = Record<string, { candidates: RawCandidate[] }>
  const raw = readJSON<RawFile>("data/results-2026.json")
  const alliances = loadAlliances()
  const names = loadAcNames()
  return Object.entries(raw)
    .map(([k, { candidates: rawC }]) => {
      const acNum = Number(k)
      const sorted = [...rawC].sort((a, b) => b.votes - a.votes)
      const winnerVotes = sorted[0]?.votes ?? 0
      const secondVotes = sorted[1]?.votes ?? 0
      const candidates: Candidate2026[] = sorted.map((c, rank) => {
        const isNota = c.name === "NOTA"
        const status: Candidate2026["status"] = isNota
          ? "nota"
          : rank === 0
            ? "won"
            : "lost"
        const margin =
          rank === 0 ? winnerVotes - secondVotes : c.votes - winnerVotes
        const alliance = isNota
          ? "NOTA"
          : ((alliances.partyToAlliance[c.party] as AllianceCode | undefined) ??
            "OTHER")
        return {
          name: c.name,
          party: c.party,
          alliance,
          votes: c.votes,
          margin,
          status,
          isNota,
        }
      })
      return {
        constituencyNumber: acNum,
        constituencyName: names[k]?.eci ?? "",
        candidates,
      }
    })
    .sort((a, b) => a.constituencyNumber - b.constituencyNumber)
}

/**
 * Load consolidated historical data. Returns Map<AC# → HistoricalConstituency>.
 */
export function loadHistorical(): Map<number, HistoricalConstituency> {
  type RawFile = Record<string, { elections: HistoricalElection[] }>
  const raw = readJSON<RawFile>("data/ac-history.json")
  const names = loadAcNames()
  const out = new Map<number, HistoricalConstituency>()
  for (const [k, v] of Object.entries(raw)) {
    const acNum = Number(k)
    out.set(acNum, {
      constituencyNumber: acNum,
      constituencyName: names[k]?.eci ?? "",
      elections: v.elections,
    })
  }
  return out
}

File: scripts/_lib/load.py
"""
Shared data loaders for python scripts.

Loads the trimmed JSON files in `data/` and rehydrates to the full
runtime shape (with derived candidate `alliance` / `status` / `margin` /
`isNota`) so analysis scripts can iterate per-candidate without
recomputing the same logic across 5+ scripts.

Mirrors `src/lib/data/constituencies.ts:hydrateConstituency` — kept
in sync by hand. The src/ runtime is the source of truth; scripts
are batch consumers.
"""
from __future__ import annotations

import json
from pathlib import Path
from typing import Any

ROOT = Path(__file__).resolve().parent.parent.parent


def _read(rel: str) -> Any:
    return json.loads((ROOT / rel).read_text())


def load_alliances() -> dict:
    return _read("data/alliances.json")


def load_ac_names() -> dict[str, dict]:
    raw = _read("data/ac-names.json")
    return {
        k: {"primary": v["primary"], "eci": v["primary"].upper()}
        for k, v in raw.items()
    }


def load_2026() -> list[dict]:
    """Return 2026 results as a list of {constituencyNumber, constituencyName,
    candidates: [{name, party, alliance, votes, margin, status, isNota}]} in
    AC order. Per-candidate alliance/status/margin/isNota are derived (the
    source JSON stores only name/party/votes — see commit ac1d440).
    """
    raw = _read("data/results-2026.json")
    alliances = load_alliances()
    pta = alliances["partyToAlliance"]
    names = load_ac_names()

    out = []
    for k, entry in raw.items():
        ac_num = int(k)
        sorted_c = sorted(entry["candidates"], key=lambda c: -c["votes"])
        winner_votes = sorted_c[0]["votes"] if sorted_c else 0
        second_votes = sorted_c[1]["votes"] if len(sorted_c) > 1 else 0
        cands = []
        for rank, c in enumerate(sorted_c):
            is_nota = c["name"] == "NOTA"
            status = "nota" if is_nota else "won" if rank == 0 else "lost"
            margin = (
                winner_votes - second_votes
                if rank == 0
                else c["votes"] - winner_votes
            )
            alliance = "NOTA" if is_nota else pta.get(c["party"], "OTHER")
            cands.append(
                {
                    "name": c["name"],
                    "party": c["party"],
                    "alliance": alliance,
                    "votes": c["votes"],
                    "margin": margin,
                    "status": status,
                    "isNota": is_nota,
                }
            )
        out.append(
            {
                "constituencyNumber": ac_num,
                "constituencyName": names.get(k, {}).get("eci", ""),
                "candidates": cands,
            }
        )
    out.sort(key=lambda c: c["constituencyNumber"])
    return out


def load_historical() -> dict[int, dict]:
    """Return {ac_num → {constituencyNumber, constituencyName, elections}}."""
    raw = _read("data/ac-history.json")
    names = load_ac_names()
    out = {}
    for k, v in raw.items():
        ac_num = int(k)
        out[ac_num] = {
            "constituencyNumber": ac_num,
            "constituencyName": names.get(k, {}).get("eci", ""),
            "elections": v["elections"],
        }
    return out


def save_json(rel_path: str, obj: Any) -> None:
    """Compact JSON write for committed data/*.json files."""
    (ROOT / rel_path).write_text(json.dumps(obj, separators=(",", ":")))
